source/beamline/BioXAS/BioXASFrontEndBeamStatusControl.cpp

```diff
@@ -0,0 +1,229 @@
+			removeChildControl(photonShutterUpstream_);
+
+		photonShutterUpstream_ = newControl;
+
+		if (photonShutterUpstream_)
+			addChildControl(photonShutterUpstream_);
+
+		emit photonShutterUpstreamChanged(photonShutterUpstream_);
+	}
+}
+
+void BioXASFrontEndBeamStatusControl::setPhotonShutterDownstream(CLSBiStateControl *newControl)
+{
+	if (photonShutterDownstream_ != newControl) {
+
+		if (photonShutterDownstream_)
+			removeChildControl(photonShutterDownstream_);
+
+		photonShutterDownstream_ = newControl;
+
+		if (photonShutterDownstream_)
+			addChildControl(photonShutterDownstream_);
+
+		emit photonShutterDownstreamChanged(photonShutterDownstream_);
+	}
+}
+
+void BioXASFrontEndBeamStatusControl::setSafetyShutter(CLSBiStateControl *newControl)
+{
+	if (safetyShutter_ != newControl) {
+
+		if (safetyShutter_)
+			removeChildControl(safetyShutter_);
+
+		safetyShutter_ = newControl;
+
+		if (safetyShutter_)
+			addChildControl(safetyShutter_);
+
+		emit safetyShutterChanged(safetyShutter_);
+	}
+}
+
+void BioXASFrontEndBeamStatusControl::setValves(BioXASValvesControl *newControl)
+{
+	if (valves_ != newControl) {
+
+		if (valves_)
+			removeChildControl(valves_);
+
+		valves_ = newControl;
+
+		if (valves_)
+			addChildControl(valves_);
+	}
+}
+
+void BioXASFrontEndBeamStatusControl::updateConnected()
+{
+	bool isConnected = (
+				photonShutterUpstream_ && photonShutterUpstream_->isConnected() &&
+				photonShutterDownstream_ && photonShutterDownstream_->isConnected() &&
+				safetyShutter_ && safetyShutter_->isConnected() &&
+				valves_ && valves_->isConnected()
+				);
+
+	setConnected(isConnected);
+}
+
+void BioXASFrontEndBeamStatusControl::updateValue()
+{
+	double value = None;
+
+	if (isConnected()) {
+
+		if (isOn())
+			value = On;
+		else if (isOff())
+			value = Off;
+		else
+			value = None;
+	}
+
+	setValue(value);
+}
+
+void BioXASFrontEndBeamStatusControl::updateMoving()
+{
+	if (isConnected()) {
+		bool isMoving = (
+					photonShutterUpstream_->isMoving() ||
+					photonShutterDownstream_->isMoving() ||
+					safetyShutter_->isMoving() ||
+					valves_->isMoving()
+					);
+
+		setIsMoving(isMoving);
+	}
+}
+
+AMAction3* BioXASFrontEndBeamStatusControl::createMoveAction(double newState)
+{
+	AMAction3 *result = 0;
+
+	switch(int(newState)) {
+	case On:
+		result = createBeamOnAction();
+		break;
+	case Off:
+		result = createBeamOffAction();
+		break;
+	default:
+		break;
+	}
+
+	return result;
+}
+
+AMAction3* BioXASFrontEndBeamStatusControl::createBeamOnAction()
+{
+	AMListAction3 *beamOn = new AMListAction3(new AMListActionInfo3("Turn beam on", "Turn beam on"), AMListAction3::Sequential);
+	beamOn->addSubAction(AMActionSupport::buildControlMoveAction(valves_, BioXASValvesControl::Open));
+	beamOn->addSubAction(AMActionSupport::buildControlMoveAction(safetyShutter_, CLSBiStateControl::Open));
+	beamOn->addSubAction(AMActionSupport::buildControlMoveAction(photonShutterDownstream_, CLSBiStateControl::Open));
```

Comment:
The downstream shutter should be the last one opened.  That is the final shutter to let light into the SOE.